File: packages/nerva-py/nerva/responder/tone.py
"""Tone responder — rewrite agent output through an LLM with tone instructions (N-614).

Applies a configurable tone (casual, professional, empathetic, etc.) to text
output via an LLM rewrite. Non-text responses (errors, empty output) pass
through without modification.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Protocol

from nerva.responder import Channel, Response
from nerva.runtime import AgentResult, AgentStatus

if TYPE_CHECKING:
    from nerva.context import ExecContext

__all__ = [
    "ToneLLMFunc",
    "ToneResponder",
    "ToneConfig",
]

# -- Constants ---------------------------------------------------------------

DEFAULT_TONE: str = "professional"
"""Default tone when none is specified."""

DEFAULT_SYSTEM_PROMPT_TEMPLATE: str = (
    "Rewrite the following text in a {tone} tone. "
    "Preserve all factual content and meaning. "
    "Return ONLY the rewritten text, nothing else."
)
"""System prompt template. ``{tone}`` is replaced with the configured tone."""


# -- LLM function protocol --------------------------------------------------


class ToneLLMFunc(Protocol):
    """Async function that sends a system+user prompt pair to an LLM.

    Args:
        system_prompt: Instructions for the LLM.
        user_prompt: The text to rewrite.

    Returns:
        Rewritten text from the LLM.
    """

    async def __call__(self, system_prompt: str, user_prompt: str) -> str: ...


# -- Config ------------------------------------------------------------------


@dataclass(frozen=True)
class ToneConfig:
    """Configuration for the tone responder.

    Attributes:
        tone: Target tone (e.g. ``"casual"``, ``"professional"``, ``"empathetic"``).
        system_prompt_template: Template string with ``{tone}`` placeholder.
    """

    tone: str = DEFAULT_TONE
    system_prompt_template: str = DEFAULT_SYSTEM_PROMPT_TEMPLATE


# -- Responder ---------------------------------------------------------------


class ToneResponder:
    """Rewrite agent output through an LLM with a configurable tone.

    Text output is sent to the LLM for rewriting. Non-text responses
    (errors, empty output) pass through without modification. Truncates
    to ``channel.max_length`` when set.

    Args:
        llm: Async LLM function for tone rewriting.
        config: Optional tone configuration overrides.
    """

    def __init__(
        self,
        llm: ToneLLMFunc,
        config: ToneConfig | None = None,
    ) -> None:
        self._llm = llm
        self._config = config or ToneConfig()

    async def format(
        self, output: AgentResult, channel: Channel, ctx: ExecContext
    ) -> Response:
        """Format agent output with tone rewriting.

        Passes through without LLM call when the output is empty or the
        agent status indicates an error.

        Args:
            output: Raw agent result from the runtime.
            channel: Target delivery channel.
            ctx: Execution context (forwarded for observability).

        Returns:
            Response with tone-rewritten text, or passthrough for non-text.
        """
        if _should_passthrough(output):
            return _passthrough_response(output, channel)

        rewritten = await self._rewrite(output.output)
        text = _apply_max_length(rewritten, channel.max_length)

        return Response(text=text, channel=channel)

    # -- Private -------------------------------------------------------------

    async def _rewrite(self, text: str) -> str:
        """Send text to the LLM for tone rewriting.

        Falls back to the original text if the LLM returns empty.

        Args:
            text: Original text to rewrite.

        Returns:
            Rewritten text, or original on failure.
        """
        system_prompt = self._config.system_prompt_template.format(
            tone=self._config.tone
        )
        try:
            result = await self._llm(system_prompt, text)
            return result.strip() if result and result.strip() else text
        except Exception:  # noqa: BLE001
            return text


# -- Pure helpers ------------------------------------------------------------


def _should_passthrough(output: AgentResult) -> bool:
    """Determine whether the output should skip tone rewriting.

    Args:
        output: Agent result to inspect.

    Returns:
        True if the output is empty or represents an error.
    """
    if output.status != AgentStatus.SUCCESS:
        return True
    if not output.output or not output.output.strip():
        return True
    return False


def _passthrough_response(output: AgentResult, channel: Channel) -> Response:
    """Build a response without tone rewriting.

    Args:
        output: Raw agent result.
        channel: Target delivery channel.

    Returns:
        Response with the raw output text.
    """
    text = output.output or output.error or ""
    text = _apply_max_length(text, channel.max_length)
    return Response(text=text, channel=channel)


def _apply_max_length(text: str, max_length: int) -> str:
    """Truncate text to max_length if the channel defines a positive limit.

    Args:
        text: Text to truncate.
        max_length: Maximum character count (0 = unlimited).

    Returns:
        Truncated text.
    """
    if max_length > 0:
        return text[:max_length]
    return text

File: packages/nerva-js/src/responder/tone.ts
/**
 * Tone responder — rewrite agent output in a configurable tone.
 *
 * Uses an LLM function to rewrite text responses while passing through
 * non-text content unchanged. Channel-aware: respects maxLength limits.
 *
 * @module responder/tone
 */

import type {
  AgentResult,
  Channel,
  ExecContext,
  Response,
  Responder,
} from "./index.js";
import { createResponse } from "./index.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/**
 * Async function that rewrites text using an LLM.
 *
 * Takes a system prompt (containing tone instructions) and the text to rewrite,
 * returns the rewritten text.
 */
export type ToneRewriteFn = (systemPrompt: string, text: string) => Promise<string>;

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/** Status values that indicate non-text / error responses (passed through). */
const PASSTHROUGH_STATUSES = new Set(["error", "timeout", "wrong_handler"]);

// ---------------------------------------------------------------------------
// ToneResponder
// ---------------------------------------------------------------------------

/**
 * Rewrites agent text output in a configurable tone via an LLM.
 *
 * Non-text responses (errors, timeouts) are passed through unchanged.
 * Applies channel maxLength truncation after rewriting.
 *
 * @example
 * ```ts
 * const responder = new ToneResponder(myLlmFn, "friendly and casual");
 * const response = await responder.format(agentResult, channel, ctx);
 * ```
 */
export class ToneResponder implements Responder {
  private readonly _rewrite: ToneRewriteFn;
  private readonly _tone: string;

  /**
   * @param rewrite - Async LLM function for tone rewriting.
   * @param tone - Description of the desired tone (e.g. "formal", "friendly and casual").
   */
  constructor(rewrite: ToneRewriteFn, tone: string) {
    this._rewrite = rewrite;
    this._tone = tone;
  }

  /**
   * Format agent output with tone rewriting.
   *
   * Non-success statuses are passed through without rewriting.
   * Empty output is returned as-is.
   *
   * @param output - Raw agent result from the runtime.
   * @param channel - Target delivery channel.
   * @param _ctx - Execution context (unused by tone responder).
   * @returns Response with tone-rewritten text.
   */
  async format(
    output: AgentResult,
    channel: Channel,
    _ctx: ExecContext,
  ): Promise<Response> {
    if (shouldPassthrough(output)) {
      return createResponse(output.output, channel);
    }

    const systemPrompt = buildTonePrompt(this._tone);
    const rewritten = await this._rewrite(systemPrompt, output.output);

    const text = applyMaxLength(rewritten, channel.maxLength);
    return createResponse(text, channel);
  }
}

// ---------------------------------------------------------------------------
// Pure helpers
// ---------------------------------------------------------------------------

/**
 * Determine if the output should be passed through without rewriting.
 *
 * @param output - Agent result to check.
 * @returns True if the output should not be rewritten.
 */
function shouldPassthrough(output: AgentResult): boolean {
  if (PASSTHROUGH_STATUSES.has(output.status)) return true;
  if (output.output === "") return true;
  return false;
}

/**
 * Build the system prompt instructing the LLM on tone.
 *
 * @param tone - Desired tone description.
 * @returns System prompt string.
 */
function buildTonePrompt(tone: string): string {
  return [
    `Rewrite the following text in a ${tone} tone.`,
    "Preserve the factual content and meaning. Return only the rewritten text.",
    "Do not add any preamble or explanation.",
  ].join(" ");
}

/**
 * Truncate text to the channel's maxLength if set.
 *
 * @param text - Text to potentially truncate.
 * @param maxLength - Maximum character length (0 = unlimited).
 * @returns Truncated text.
 */
function applyMaxLength(text: string, maxLength: number): string {
  if (maxLength > 0) {
    return text.slice(0, maxLength);
  }
  return text;
}
